tripal/tripal
# Improving tripal docker performance

# Tripal 4 Core Dev Task

### Issue #1857


### Tripal Version: 4

## Description


The first page load after starting a new docker is very slow. This is only intensified as more modules are added to the site. Some Drupal issues have claimed this may due to "all those thousands of Drupal php files have to be loaded into the opcache before it runs" and "class loading performance". This is causing full on page timeouts on some of our extension development dockers based on Tripaldocker and thus needs to be addressed 🙈

### NOTE: Not complete as there is not a substantial enough improvement yet

### What this PR does

- Configure supervisord to also start postgresql when the container is run!
- Add APCu caching to the docker to replace the slower symphony class loader.
- Fix pear configuration issue where the path to the ini was not set.
- Add [Webprofiler Toolbar](https://www.drupal.org/project/webprofiler) module to our docker containers so we can actually see performance better ;-p

### Webprofiler

Adds a teeny-tiny toolbar to the bottom of all Drupal pages indicating helpful stats like the number of queries, total page time, total memory used, ect. (See Screenshot)

<img width="1265" alt="Screenshot 2024-04-23 at 5 17 59 PM" src="https://github.com/tripal/tripal/assets/1566301/adaeb26e-1fec-4ab8-a66b-3f2f619990dd">


## Testing?




This is a VERY hard one to test/confirm. The main bottleneck is the first page load after creating a container. As such it can only be tested once per container 🤪 

1. Create a container NOT ON this branch. **Do NOT load the container in the browser yet!** Follow the instructions below to check first page load.
2. Switch to this branch and create a new container. **Do NOT load the container in the browser yet!** Follow the instructions below to check first page load.

NOTE: each time you want to test the first page load you have to start a fresh container! subsequent refreshes do not show the same issue.

### Notes

These will be populated once I figure out a good way to time this 🤪 Likely using our built-in xdebug profiling.

**Do NOT load the container in the browser yet! Wait until instructed to do so!**

1. Turn on Xdebug profiling in your container. Replace CONTAINERNAME with the name of your container in the following command: `docker exec --workdir=/var/www/drupal/web/modules/contrib/tripal CONTAINERNAME xdebug_toggle.sh`.
2. Check the settings using drush to confirm it was enabled:
```
docker exec tripal1858 drush php:eval "xdebug_info();" | grep "rofiler"
Profiler => ✔ enabled
                                 Profiler                                 
Profiler is not active
xdebug.profiler_output_name => cachegrind.out.%p => cachegrind.out.%p
xdebug.profiler_append => Off => Off
```
3. Create the directory which will hold the profiler logs: `docker exec --workdir=/var/www/drupal/web/modules/contrib/tripal CONTAINERNAME mkdir tripaldocker/xdebug_output`
4. Now, in your browse go to http://localhost/?XDEBUG_TRIGGER=1. This will load the frontpage and also trigger profiling. Once the page has fully loaded you should see a file in the directory `/var/www/drupal/web/modules/contrib/tripal/tripaldocker/xdebug_output`.
5. To visualize we will use webgrind.
